### PR TITLE
Get Rid of nobuild and nolink

### PR DESCRIPTION
**Pull request by citibeth**
The dependency type aliases `nobuild` and `nolink` have been deprecated.  This PR gets rid of them.

@tgamblin @hegner Would you be able to please take a quick look at the chnages to `database.py`?  I think they need a little help...

### CODE REVIEW

**Comment by citibeth** [comment]
Actually, I think that `database.py` is OK.  As long as `_tracked_deps` can be a tuple in the call to `rec.spec.dependencies(_tracked_deps)`.

**Comment by citibeth** [comment]
@adamjstewart Do you know why Travis tests have stalled here (or how to restart)?  Thanks!

**Comment by adamjstewart** [comment]
Not sure why they stalled. But yes, pushing a new commit is one way to get them to restart. So is closing and reopening the PR. I think if you go to the travis page linked to under details, there is sometimes a restart button as well.

**Comment by citibeth** [comment]
Hmm... pushing the new commit didn't work.  Can't find a restart button.  now trying closing/reopening.

**Comment by citibeth** [comment]
Still no tests :*(

**Comment by adamjstewart** [comment]
Travis definitely seems less stable now that we switched to pytest and added macOS tests. @alalazo @tgamblin?

**Comment by citibeth** [comment]
This passed all the tests; except Travis hung and hasn't yet started the Mac tests.  I'm adding the "ready" label...

**Comment by tgamblin** [comment]
It's not hung -- it's just waiting for the mac worker.  I suspect those are scarce, but I'd rather have mac tests than not.

**Comment by adamjstewart** [comment]
The mac tests take quite a while. If you check in `.travis.yml`, you'll see things like `brew update`. That's what causes the tests to take so long.

**Comment by adamjstewart** [comment]
Ok, so it takes a while to find an available mac worker, and the tests take a while to run `brew update`. That makes more sense.

**Comment by citibeth** [comment]
I understand it may take a while... I'm still betting optimistically that
the test will succeed.

**Comment by adamjstewart** [comment]
P.S. This may conflict a lot with #2761.

**Comment by citibeth** [comment]
They modify different lines; so I think both PRs will go in with few or no
conflicts.
But if there are conflicts... we'll figure it out then.

File: lib/spack/docs/packaging_guide.rst [hunk 1]
```diff
@@ -1508,10 +1508,7 @@ Additional hybrid dependency types are (note the lack of quotes):
   * **<not specified>**: ``type`` assumed to be ``("build",
     "link")``. This is the common case for compiled language usage.
   * **alldeps**: All dependency types.  **Note:** No quotes here
```
Comment (tgamblin) [inline lib/spack/docs/packaging_guide.rst]: `alldeps` can probably be eliminated from the user docs and packages, as well.  It's used internally by the Spec traversal routines, which need a way to guarantee that they are going over the *whole* DAG.  

Using `alldeps` in a package is going be very surprising when we add things like `test` dependencies.  There are 4 packages that currently use `alldeps` -- those should probably be changed to `('build', 'link', 'run')`.

**Comment by citibeth** [comment]
Closing; the SHA Travis don't match the SHA on the source branch (`5882cdc8`), hence the errors:
https://github.com/citibeth/spack/commits/efischer/170105-EliminateNolink

**Comment by tgamblin** [comment]
@citibeth: I think you are seeing the effect of Travis merging the PR into the target branch before building.  If you [go here](https://travis-ci.org/LLNL/spack/pull_requests) and click on the SHA you see there, it takes you to a page [like this one](https://github.com/LLNL/spack/commit/ec1c44fb337e2d658de54c24d0d2e82c7dada04a).  You see there that the travis commit is ec1c44fb337e2d658de54c24d0d2e82c7dada04a, which is a merge of 5882cdc843d00f05508e2c9b10cca45f9a352d5e into the tip of `develop`.

**Comment by citibeth** [comment]
@tgamblin  OK, looks like it's OK now.  My trouble was actually rather prosaic: I had neglected to commit the most recent change from my working tree.  So it was not in Git, and Travis could not see it.  And no amount of pushing and re-pushing my branch could change that.

**Comment by citibeth** [comment]
@tgamblin I have addressed your change request.

**Pull request merged**